From aac88b3d4d772c19140f8e2f2f6fd0b8243ea201 Mon Sep 17 00:00:00 2001
From: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Date: Mon, 26 Aug 2019 09:41:57 -0700
Subject: [PATCH 0663/1852] Revert "Revert "dma: xilinx: Release buffers before
 DMA transfer""

commit ab964691071e95e980c2709ae584a929258d5034 from
https://github.com/Xilinx/linux-xlnx.git

We need the early callback for low latency to return buffer at
start of descriptor processing as without it there is no way for
dma client and hence also the application to be aware of
actual position of dma.

So without earlycallback, it leads to a problem as buffers
carry the timestamps based upon when they were queued and not
when dma has actually started writing onto it.

This reverts commit 4036801ee0c1ceef86bbe8dd2f3a39408b224161.

Signed-off-by: Devarsh Thakkar <devarsh.thakkar@xilinx.com>
Reviewed-by: Hyun Kwon <hyun.kwon@xilinx.com>
State: pending
Signed-off-by: Yaliang Wang <Yaliang.Wang@windriver.com>
---
 drivers/dma/xilinx/xilinx_frmbuf.c | 13 +++++++++++++
 include/linux/dma/xilinx_frmbuf.h  |  1 +
 2 files changed, 14 insertions(+)

diff --git a/drivers/dma/xilinx/xilinx_frmbuf.c b/drivers/dma/xilinx/xilinx_frmbuf.c
index 2545063c9b70..a135af17129b 100644
--- a/drivers/dma/xilinx/xilinx_frmbuf.c
+++ b/drivers/dma/xilinx/xilinx_frmbuf.c
@@ -1096,6 +1096,19 @@ static void xilinx_frmbuf_start_transfer(struct xilinx_frmbuf_chan *chan)
 				struct xilinx_frmbuf_tx_descriptor,
 				node);
 
+	if (desc->earlycb == EARLY_CALLBACK_LOW_LATENCY) {
+		dma_async_tx_callback callback;
+		void *callback_param;
+
+		callback = desc->async_tx.callback;
+		callback_param = desc->async_tx.callback_param;
+		if (callback) {
+			callback(callback_param);
+			desc->async_tx.callback = NULL;
+			chan->active_desc = desc;
+		}
+	}
+
 	/* Start the transfer */
 	chan->write_addr(chan, XILINX_FRMBUF_ADDR_OFFSET,
 			 desc->hw.luma_plane_addr);
diff --git a/include/linux/dma/xilinx_frmbuf.h b/include/linux/dma/xilinx_frmbuf.h
index 739ccccfcac4..5307113efd5b 100644
--- a/include/linux/dma/xilinx_frmbuf.h
+++ b/include/linux/dma/xilinx_frmbuf.h
@@ -16,6 +16,7 @@
 
 /* Modes to enable early callback */
 #define EARLY_CALLBACK			BIT(1) /* To avoid first frame delay */
+#define EARLY_CALLBACK_LOW_LATENCY	BIT(2) /* Low latency mode */
 
 /**
  * enum vid_frmwork_type - Linux video framework type
-- 
2.31.1

